test/e2e/upgrade_cluster_test.go

```diff
@@ -0,0 +1,189 @@
+//go:build e2e
+// +build e2e
+
+package e2e
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"testing"
+
+	"github.com/onsi/gomega"
+	hyperv1 "github.com/openshift/hypershift/api/hypershift/v1beta1"
+	e2eutil "github.com/openshift/hypershift/test/e2e/util"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	nodePoolAnnotationCurrentConfig        = "hypershift.openshift.io/nodePoolCurrentConfig"
+	nodePoolAnnotationCurrentConfigVersion = "hypershift.openshift.io/nodePoolCurrentConfigVersion"
+	hostedClusterUpgradeTestLabel          = "hypershift.openshift.io/upgrade-test"
+)
+
+// TestUpgradeHyperShift validates that a HyperShift Operator upgrade won't unnecessarily cause a node rollout.
+// The test must be triggered via the flag upgrade.run-tests and the HyperShift Operator initial image must be set
+// via the flag e2e.hypershift-operator-initial-image.
+func TestUpgradeHyperShift(t *testing.T) {
+	if globalOpts.RunUpgradeTest {
+		var once sync.Once
+		var cleanup e2eutil.Cleanup
+		var err error
+
+		var mgmtClient crclient.Client
+		var hostedCluster *hyperv1.HostedCluster
+
+		var nodePoolsMap map[string]*hyperv1.NodePool
+
+		hyperShiftOperatorLatestImage := globalOpts.HyperShiftOperatorLatestImage
+		hyperShiftOperatorInitialImage := globalOpts.HyperShiftOperatorInitialImage
+
+		t.Parallel()
+		ctx, cancel := context.WithCancel(testContext)
+		defer cancel()
+
+		clusterOpts := globalOpts.DefaultClusterOptions(t)
+
+		t.Log("Starting HyperShift Operator upgrade test")
+		zones := strings.Split(globalOpts.ConfigurableClusterOptions.Zone.String(), ",")
+		if len(zones) >= 3 {
+			t.Logf("Sufficient zones available for InfrastructureAvailabilityPolicy HighlyAvailable")
+			clusterOpts.AWSPlatform.Zones = zones
+			clusterOpts.InfrastructureAvailabilityPolicy = string(hyperv1.HighlyAvailable)
+			clusterOpts.NodePoolReplicas = 1
+		}
+
+		clusterOpts.BeforeApply = func(o crclient.Object) {
+			switch obj := o.(type) {
+			case *hyperv1.HostedCluster:
+				if obj.Annotations == nil {
+					obj.Annotations = make(map[string]string)
+				}
+				obj.Annotations[hyperv1.ControlPlaneOperatorV2Annotation] = "true"
+
+				// Add a label to identify the HostedCluster for upgrade tests in case they're leaked
+				// and need to be cleaned up.
+				if obj.Labels == nil {
+					obj.Labels = make(map[string]string)
+				}
+				obj.Labels[hostedClusterUpgradeTestLabel] = "true"
+			}
+		}
+
+		t.Run("Install initial HyperShift Operator", func(tt *testing.T) {
+			tt.Logf("Installing HyperShift Operator with image %s", hyperShiftOperatorInitialImage)
+			globalOpts.HyperShiftOperatorLatestImage = hyperShiftOperatorInitialImage
+			cleanup, err = e2eutil.InstallHyperShiftOperatorFull(tt, ctx, globalOpts)
+			if err != nil {
+				tt.Error("Failed to install initial HyperShift Operator")
+				tt.Fail()
+				t.FailNow()
+			}
+			// Set up the cleanup function on the outer test
+			t.Cleanup(func() {
```

Comment:
Registering cleanup for the outer test from a subtest is weird. Why are we doing that?

Reply:
Can you clarify why is it weird? I think weird is pretty subjective :) .

Reply:
imo, I don't think 'install hypershit operator' should be a subtest, its should just be a part of the main test, which will eliminate this problem and the extra checks.

Reply:
Addressed.